models/migrations/v111.go

```diff
@@ -5,35 +5,373 @@
 package migrations
 
 import (
-	"code.gitea.io/gitea/models"
+	"fmt"
 
 	"xorm.io/xorm"
 )
 
 func addBranchProtectionCanPushAndEnableWhitelist(x *xorm.Engine) error {
-
 	type ProtectedBranch struct {
-		CanPush                  bool  `xorm:"NOT NULL DEFAULT false"`
-		EnableApprovalsWhitelist bool  `xorm:"NOT NULL DEFAULT false"`
-		RequiredApprovals        int64 `xorm:"NOT NULL DEFAULT 0"`
+		CanPush                   bool    `xorm:"NOT NULL DEFAULT false"`
+		EnableApprovalsWhitelist  bool    `xorm:"NOT NULL DEFAULT false"`
+		ApprovalsWhitelistUserIDs []int64 `xorm:"JSON TEXT"`
+		ApprovalsWhitelistTeamIDs []int64 `xorm:"JSON TEXT"`
+		RequiredApprovals         int64   `xorm:"NOT NULL DEFAULT 0"`
+	}
+
+	type User struct {
+		ID   int64 `xorm:"pk autoincr"`
+		Type int
+
+		// Permissions
+		IsAdmin      bool
+		IsRestricted bool `xorm:"NOT NULL DEFAULT false"`
+		Visibility   int  `xorm:"NOT NULL DEFAULT 0"`
 	}
 
 	type Review struct {
 		ID       int64 `xorm:"pk autoincr"`
 		Official bool  `xorm:"NOT NULL DEFAULT false"`
-	}
 
-	sess := x.NewSession()
-	defer sess.Close()
+		ReviewerID int64 `xorm:"index"`
+		IssueID    int64 `xorm:"index"`
+	}
 
-	if err := sess.Sync2(new(ProtectedBranch)); err != nil {
+	if err := x.Sync2(new(ProtectedBranch)); err != nil {
```

Comment:
prevent sqlite deadlocks

from time to time if you Alter a table in sqlite you get locks when writting stuff to it